pkg/operator/staleconditions/remove_stale_conditions.go

```diff
@@ -63,55 +42,3 @@ func (c RemoveStaleConditions) sync() error {
 
 	return nil
 }
-
-// Run starts the kube-scheduler and blocks until stopCh is closed.
-func (c *RemoveStaleConditions) Run(ctx context.Context, workers int) {
-	defer utilruntime.HandleCrash()
-	defer c.queue.ShutDown()
-
-	klog.Infof("Starting RemoveStaleConditions")
-	defer klog.Infof("Shutting down RemoveStaleConditions")
-
-	if !cache.WaitForCacheSync(ctx.Done(), c.cachesToSync...) {
-		utilruntime.HandleError(fmt.Errorf("caches did not sync"))
-		return
-	}
-
-	// doesn't matter what workers say, only start one.
-	go wait.Until(c.runWorker, time.Second, ctx.Done())
-
-	<-ctx.Done()
-}
-
-func (c *RemoveStaleConditions) runWorker() {
-	for c.processNextWorkItem() {
-	}
-}
-
-func (c *RemoveStaleConditions) processNextWorkItem() bool {
-	dsKey, quit := c.queue.Get()
-	if quit {
-		return false
-	}
-	defer c.queue.Done(dsKey)
-
-	err := c.sync()
-	if err == nil {
-		c.queue.Forget(dsKey)
-		return true
-	}
-
-	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", dsKey, err))
-	c.queue.AddRateLimited(dsKey)
-
-	return true
-}
-
-// eventHandler queues the operator to check spec and status
-func (c *RemoveStaleConditions) eventHandler() cache.ResourceEventHandler {
-	return cache.ResourceEventHandlerFuncs{
-		AddFunc:    func(obj interface{}) { c.queue.Add(workQueueKey) },
-		UpdateFunc: func(old, new interface{}) { c.queue.Add(workQueueKey) },
-		DeleteFunc: func(obj interface{}) { c.queue.Add(workQueueKey) },
```

Comment:
as far as I see the key names are out of control of the controller authors. That's bad. This is a core decission what to queue when. In the base controller we queue `name+"Key"` while other controllers want actual object names to be queued. That makes a huge difference in the context of multiple works, but also in the context of `c.queue.Forget/Done`.

Reply:
@sttts I made `WithQueueKeyFunc` that allows you to specify custom queue key function. By default I use the one we already had (based on controller name). I'm open to add more extensibility if we need to (like add map[informer]keyFunc or something).